File: python/benchmark/bench_lsh.py
import argparse
import json
import os
import sys
import time
from pathlib import Path

import numpy as np

from lsh import RandomProjectionLSH


def read_fvecs(fp):
    """Read fvecs file into numpy array"""
    a = np.fromfile(fp, dtype="int32")
    d = a[0]
    return a.reshape(-1, d + 1)[:, 1:].copy().view("float32")


def get_gt_top_k_indices(q, X, k):
    """Get top k indices of ground truth vectors by cosine similarity"""
    q_norm = q / np.linalg.norm(q)
    X_norm = X / np.linalg.norm(X, axis=1, keepdims=True)
    cos_sims = np.dot(X_norm, q_norm)
    return np.argsort(cos_sims)[-k:][::-1]


def compute_recall(lsh_indices, gt_top_k_indices):
    """Calculate recall score"""
    if len(gt_top_k_indices) == 0:
        return 0.0
    lsh_set = set(lsh_indices)
    gt_set = set(gt_top_k_indices)
    intersection = lsh_set.intersection(gt_set)
    return len(intersection) / len(gt_set)


def evaluate_recall(Q, X, all_neighbors, n_eval_queries, verbose=True):
    """Evaluate recall for the first n_eval_queries queries"""
    per_query_recall = []
    queries_with_candidates = 0

    for i in range(n_eval_queries):
        lsh_indices = all_neighbors[i]
        k = len(lsh_indices)

        if k == 0:
            per_query_recall.append(0.0)
            if verbose:
                print(f"  Query {i}: no candidates")
            continue

        queries_with_candidates += 1
        gt_indices = get_gt_top_k_indices(Q[i], X, k)
        recall_score = compute_recall(lsh_indices, gt_indices)
        per_query_recall.append(recall_score)

        if verbose:
            print(f"  Query {i}: recall={recall_score:.4f} (k={k})")

    avg_recall = np.mean(per_query_recall) if per_query_recall else 0.0

    return {
        "n_eval_queries": n_eval_queries,
        "queries_with_candidates": queries_with_candidates,
        "avg_recall": float(avg_recall),
        "per_query_recall": [float(r) for r in per_query_recall],
    }


def run_benchmark():
    parser = argparse.ArgumentParser(
        description="Benchmark pure-Python LSH on SIFT dataset"
    )
    parser.add_argument("-d", "--data-dir", type=str, required=True)
    parser.add_argument("-nh", "--n-hash-tables", type=int, default=16)
    parser.add_argument("-np", "--n-projections", type=int, default=4)
    parser.add_argument("-s", "--seed", type=int, default=42)
    parser.add_argument("-nq", "--n-queries", type=int, default=100)
    parser.add_argument("-ne", "--n-eval-queries", type=int, default=10)
    parser.add_argument("-r", "--results-dir", type=str, default=None)
    parser.add_argument("-o", "--save-dir", type=str, default=None)
    args = parser.parse_args()

    data_dir = Path(args.data_dir)
    if not data_dir.exists():
        raise FileNotFoundError(
            f"Data dir '{data_dir}' not found. Run download_sift1m first"
        )

    X = read_fvecs(data_dir / "sift_base.fvecs")
    Q = read_fvecs(data_dir / "sift_query.fvecs")

    print(f"Data shape: {X.shape}")
    print(f"Query shape: {Q.shape}")

    Q_test = Q[: args.n_queries]
    print(f"Using {len(Q_test)} test queries")

    lsh = RandomProjectionLSH(
        n_hash_tables=args.n_hash_tables,
        n_projections=args.n_projections,
        seed=args.seed,
    )

    print()
    print("LSH Model:")
    print(f"  n_hash_tables: {lsh.n_hash_tables}")
    print(f"  n_projections: {lsh.n_projections}")
    print(f"  seed: {lsh.seed}")
    print()

    try:
        # Fit
        print("Running fit()...")
        start_time = time.time()
        model = lsh.fit(X)
        fit_time = time.time() - start_time
        print(f"fit() completed in {fit_time:.2f}s")
        print()

        # Query
        print("Running query()...")
        start_time = time.time()
        all_neighbors = model.query(Q_test)
        query_time = time.time() - start_time
        print(f"query() completed in {query_time:.2f}s")
        total_candidates = sum(len(n) for n in all_neighbors)
        print(f"Total candidates: {total_candidates}")
        print()

        # Evaluate recall
        n_eval = min(args.n_eval_queries, len(Q_test))
        print(f"Evaluating recall on first {n_eval} queries:")
        recall_results = evaluate_recall(Q_test, X, all_neighbors, n_eval, verbose=True)
        print()
        print(f"Average recall: {recall_results['avg_recall']:.4f}")
        print(f"Queries with candidates: {recall_results['queries_with_candidates']}/{n_eval}")
        print()

        # Save report
        if args.results_dir:
            report_path = f"{args.results_dir}/report_h{args.n_hash_tables}_p{args.n_projections}_{time.strftime('%Y%m%d_%H%M%S')}.json"
            os.makedirs(args.results_dir, exist_ok=True)

            with open(report_path, "w") as f:
                report_data = {
                    "params": {
                        "n_hash_tables": args.n_hash_tables,
                        "n_projections": args.n_projections,
                        "seed": args.seed,
                        "n_queries": args.n_queries,
                    },
                    "runtimes": {
                        "fit_time": fit_time,
                        "query_time": query_time,
                    },
                    "stats": {
                        "total_candidates": total_candidates,
                    },
                    "recall_evaluation": recall_results,
                }
                json.dump(report_data, f, indent=4)

            print(f"Report saved to {report_path}")
        else:
            print("Runtimes:")
            print(f"  fit_time: {fit_time:.2f}s")
            print(f"  query_time: {query_time:.2f}s")

        # Save model if requested
        if args.save_dir:
            print(f"Saving model to {args.save_dir}...")
            model.save(args.save_dir)
            print("Model saved successfully")

    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    run_benchmark()

File: cpp/benchmark/bench_lsh.cpp
#include <random_projection_lsh.h>

#include <Eigen/Dense>
#include <chrono>
#include <filesystem>
#include <fstream>
#include <getopt.h>
#include <iomanip>
#include <iostream>
#include <set>
#include <sstream>
#include <string>
#include <vector>

using namespace Eigen;
using namespace std;
namespace fs = filesystem;

/*
 * Read SIFT .fvecs file into MatrixXd
 */
MatrixXd read_fvecs(const string& filepath) {
    ifstream file(filepath, ios::binary);
    if (!file.is_open()) {
        throw runtime_error("Failed to open file '" + filepath + "'");
    }

    // read input dimension
    int d;
    file.read(reinterpret_cast<char*>(&d), sizeof(int));

    file.seekg(0, ios::end);
    streamsize file_size = file.tellg();
    file.seekg(0, ios::beg);

    int vector_size = sizeof(int) + d * sizeof(float);
    int n_vectors = file_size / vector_size;

    MatrixXd data(n_vectors, d);

    // read all vectors into MatrixXd
    for (int i = 0; i < n_vectors; ++i) {
        int dim;
        file.read(reinterpret_cast<char*>(&dim), sizeof(int));
        if (dim != d) {
            throw runtime_error("Inconsistent dimensions in file");
        }

        vector<float> vec(d);
        file.read(reinterpret_cast<char*>(vec.data()), d * sizeof(float));

        for (int j = 0; j < d; ++j) {
            data(i, j) = static_cast<double>(vec[j]);
        }
    }

    return data;
}

vector<int> get_gt_top_k_indices(const VectorXd& q, const MatrixXd& X, int k) {
    VectorXd q_norm = q / q.norm();
    VectorXd cos_sims(X.rows());
    for (MatrixXd::Index i = 0; i < X.rows(); ++i) {
        VectorXd x_norm = X.row(i) / X.row(i).norm();
        cos_sims(i) = x_norm.dot(q_norm);
    }

    vector<pair<double, int>> sim_idx;
    for (int i = 0; i < cos_sims.rows(); ++i) {
        sim_idx.push_back({cos_sims(i), i});
    }

    sort(sim_idx.begin(), sim_idx.end(), greater<pair<double, int>>());

    vector<int> top_k;
    for (int i = 0; i < min(k, static_cast<int>(sim_idx.size())); ++i) {
        top_k.push_back(sim_idx[i].second);
    }

    return top_k;
}

double calculate_recall(const vector<int>& lsh_indices, const vector<int>& gt_indices) {
    set<int> lsh_set(lsh_indices.begin(), lsh_indices.end());
    set<int> gt_set(gt_indices.begin(), gt_indices.end());

    set<int> intersection;
    set_intersection(lsh_set.begin(), lsh_set.end(), gt_set.begin(), gt_set.end(),
                     inserter(intersection, intersection.begin()));

    return static_cast<double>(intersection.size()) / gt_set.size();
}

struct Config {
    fs::path data_dir;
    int n_hash_tables = 16;
    int n_projections = 4;
    unsigned int seed = random_device{}();
    int n_queries = 100;
    fs::path save_dir = "";
};

void print_usage(const char* program_name) {
    cout << "Usage: " << program_name << " [options]\n"
         << "Options:\n"
         << "  -d, --data-dir        Data directory (default: data/sift)\n"
         << "  -h, --n-hash-tables   Number of hash tables (default: 16)\n"
         << "  -p, --n-projections   Number of projections per table (default: 4)\n"
         << "  -s, --seed            Random seed (default: random)\n"
         << "  -q, --num-queries     Number of test queries (default: 100)\n"
         << "  -o, --save-dir        Directory to save model (optional)\n";
}

Config parse_args(int argc, char* argv[]) {
    Config conf;

    static struct option long_options[] = {
        {"data-dir", required_argument, 0, 'd'},      {"n-hash-tables", required_argument, 0, 'h'},
        {"n-projections", required_argument, 0, 'p'}, {"seed", required_argument, 0, 's'},
        {"num-queries", required_argument, 0, 'q'},   {"save-dir", required_argument, 0, 'o'},
    };

    int c;
    while ((c = getopt_long(argc, argv, "d:h:p:s:q:o:", long_options, nullptr)) != -1) {
        switch (c) {
        case 'd':
            conf.data_dir = optarg;
            break;
        case 'h':
            conf.n_hash_tables = atoi(optarg);
            break;
        case 'p':
            conf.n_projections = atoi(optarg);
            break;
        case 's':
            conf.seed = atoi(optarg);
            break;
        case 'q':
            conf.n_queries = atoi(optarg);
            break;
        case 'o':
            conf.save_dir = optarg;
            break;
        default:
            print_usage(argv[0]);
            exit(1);
        }
    }

    return conf;
}

int main(int argc, char* argv[]) {
    Config conf = parse_args(argc, argv);

    if (!fs::exists(conf.data_dir) || !fs::is_directory(conf.data_dir)) {
        throw runtime_error("Directory '" + conf.data_dir.string() + "' not found");
    }

    MatrixXd X = read_fvecs(conf.data_dir / "sift_base.fvecs");
    MatrixXd Q = read_fvecs(conf.data_dir / "sift_query.fvecs");

    cout << "Data shape: " << X.rows() << "x" << X.cols() << endl;
    cout << "Query shape: " << Q.rows() << "x" << Q.cols() << endl;

    auto Q_test = Q.topRows(conf.n_queries);
    cout << "Using " << conf.n_queries << " test queries" << endl;

    // create model
    RandomProjectionLSH lsh(conf.n_hash_tables, conf.n_projections, conf.seed);

    cout << "\nCreated LSH Model:" << endl;
    cout << "  n_hash_tables: " << lsh.get_n_hash_tables() << endl;
    cout << "  n_projections: " << lsh.get_n_projections() << endl;
    cout << "  seed: " << lsh.get_seed() << endl;
    cout << endl;

    // fit
    cout << "Running fit()..." << endl;
    auto start_time = chrono::high_resolution_clock::now();
    auto model = lsh.fit(X);
    auto fit_time = chrono::high_resolution_clock::now() - start_time;
    auto fit_seconds = chrono::duration_cast<chrono::duration<double>>(fit_time).count();
    cout << "fit() completed in " << fit_seconds << "s" << endl << endl;

    // query
    cout << "Running query()..." << endl;
    start_time = chrono::high_resolution_clock::now();
    auto all_neighbors = model.query_indices(Q_test);
    auto query_time = chrono::high_resolution_clock::now() - start_time;
    auto query_seconds = chrono::duration_cast<chrono::duration<double>>(query_time).count();
    cout << "query() completed in " << query_seconds << "s" << endl << endl;

    // calculate recall for first query
    double recall_score = 0.0;
    int intersection_size = 0;
    int gt_size = 0;

    if (!all_neighbors.empty() && !all_neighbors[0].empty()) {
        vector<int> lsh_indices = all_neighbors[0];
        gt_size = lsh_indices.size();
        vector<int> gt_indices = get_gt_top_k_indices(Q_test.row(0), X, gt_size);
        recall_score = calculate_recall(lsh_indices, gt_indices);

        set<int> lsh_set(lsh_indices.begin(), lsh_indices.end());
        set<int> gt_set(gt_indices.begin(), gt_indices.end());
        set<int> intersection;
        set_intersection(lsh_set.begin(), lsh_set.end(), gt_set.begin(), gt_set.end(),
                         inserter(intersection, intersection.begin()));
        intersection_size = intersection.size();

        cout << "First query recall: " << recall_score << " (" << intersection_size << "/"
             << gt_size << ")" << endl;
    } else {
        cout << "No candidates found for first query" << endl;
    }

    // save report
    if (!fs::create_directories("results") && !fs::exists("results")) {
        throw runtime_error("Failed to create results directory");
    }
    auto now = chrono::system_clock::now();
    auto time_t = chrono::system_clock::to_time_t(now);
    stringstream ss;
    ss << put_time(localtime(&time_t), "%Y%m%d_%H%M%S");
    string report_path = "results/report_h" + to_string(conf.n_hash_tables) + "_p" + to_string(conf.n_projections) + "_" + ss.str() + ".json";

    ofstream report(report_path);
    report << "{\n";
    report << "    \"params\": {\n";
    report << "        \"n_hash_tables\": " << conf.n_hash_tables << ",\n";
    report << "        \"n_projections\": " << conf.n_projections << ",\n";
    report << "        \"seed\": " << lsh.get_seed() << ",\n";
    report << "        \"num_queries\": " << conf.n_queries << "\n";
    report << "    },\n";
    report << "    \"runtimes\": {\n";
    report << "        \"fit_time\": " << fit_seconds << ",\n";
    report << "        \"query_time\": " << query_seconds << "\n";
    report << "    },\n";
    report << "    \"first_query_results\": {\n";
    report << "        \"recall_score\": " << recall_score << ",\n";
    report << "        \"intersection_size\": " << intersection_size << ",\n";
    report << "        \"gt_size\": " << gt_size << "\n";
    report << "    }\n";
    report << "}\n";
    report.close();

    cout << "Report saved to " << report_path << endl;

    // optionally save model
    if (!conf.save_dir.empty()) {
        cout << "Saving model to " << conf.save_dir << "..." << endl;
        model.save(conf.save_dir);
        cout << "Model saved successfully" << endl;
    }
}
